src/async/smtp/MCSMTPSendWithDataOperation.cpp

```diff
@@ -84,3 +84,9 @@ void SMTPSendWithDataOperation::main()
     }
     setError(error);
 }
+
+void SMTPSendWithDataOperation::cancel()
+{
+    SMTPOperation::cancel();
+    session()->session()->cancelMessageSending();
```

Comment:
Could you get rid of setSendingCancelled()?
and check in main() if the operation is cancelled before running sendMessage()?